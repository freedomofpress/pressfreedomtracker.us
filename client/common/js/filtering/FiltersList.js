import React, { PureComponent } from 'react'
import moment from 'moment'
import {
	AUTOCOMPLETE_SINGLE_FILTERS,
	AUTOCOMPLETE_MULTI_FILTERS,
	DATE_FILTERS,
	HUMAN_DATE_FORMAT,
} from '~/filtering/constants'


class FiltersList extends PureComponent {
	renderFilterValue(filter, filterValues) {
		let title = filter.title
		let renderedValue = filterValues[filter.name]
		if (filter.type === 'date') {
			title = title.slice(0, -8)
			const lowerDate = filterValues[`${filter.name}_lower`]
			const upperDate = filterValues[`${filter.name}_upper`]

			if (upperDate && lowerDate) {
				title = `${title} between`
				renderedValue = `${moment(lowerDate).format(HUMAN_DATE_FORMAT)} â€“ ${moment(upperDate).format(HUMAN_DATE_FORMAT)}`
			} else if (upperDate) {
				title = `${title} before`
				renderedValue = moment(upperDate).format(HUMAN_DATE_FORMAT)
			} else {
				title = `${title} since`
				renderedValue = moment(lowerDate).format(HUMAN_DATE_FORMAT)
			}
		} else if (filter.type === 'boolean') {
			renderedValue = filterValues[filter.name] ? 'yes' : 'no'
		} else if (filter.type === 'choice') {
			const selected = filter.choices.find(choice => choice[0] === filterValues[filter.name])
			if (selected) renderedValue = selected[1]
		} else if (filter.type === 'autocomplete') {
			if (Array.isArray(filterValues[filter.name])) {
				renderedValue = (
					<ul className="filters__summary-list">
						{filterValues[filter.name].map(({ label, id }) => (
							<li
								key={id}
								className="filters__summary-item"
							>
								{label}
							</li>
						))}
					</ul>
				)
			} else {
				renderedValue = filterValues[filter.name].label
			}
		}

		return (
			<span>
				<span className="filters__text--dim">{title}:</span>
				{' '}
				{renderedValue}
			</span>
		)
	}

	render() {
		const {
			categories,
			categoriesEnabled,
			filterValues,
		} = this.props

		const filters = categories.filter(({ id }) => categoriesEnabled[id]).reduce((acc, category) => {
			return [
				...acc,
				...category.filters.filter(filter => {
					if (filter.type === 'date') {
						return filterValues[`${filter.name}_lower`] || filterValues[`${filter.name}_upper`]
					} else {
						return filterValues[filter.name]
					}
				}),
			]
		}, [])

		return (
			<ul className="filters__summary-list">
				{filters.map(filter => (
					<li key={filter.name} className="filters__summary-item">
						{this.renderFilterValue(filter, filterValues)}
					</li>
				))}
			</ul>
		)
	}
}


export default FiltersList
