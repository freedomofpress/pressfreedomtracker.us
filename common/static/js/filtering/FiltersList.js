import React, { PureComponent } from 'react'
import {
	AUTOCOMPLETE_SINGLE_FILTERS,
	AUTOCOMPLETE_MULTI_FILTERS,
	DATE_FILTERS,
	HUMAN_DATE_FORMAT,
} from '~/filtering/constants'


class FiltersList extends PureComponent {
	renderValue(label, value) {
		if (DATE_FILTERS.includes(label)) {
			var formattedValue = value.format(HUMAN_DATE_FORMAT)
		} else {
			var formattedValue = value
		}

		const underscoreRe = /_/g

		if (DATE_FILTERS.includes(label)) {
			const period = label.includes('_lower') ? 'since' : 'before'
			const description = label.substring(0, label.length - 6).replace(underscoreRe, ' ')
			return (
				<span>
					<span className="filters__text--dim">{description} {period}</span>
					{' '}
					{formattedValue}
				</span>
			)
		} else if (AUTOCOMPLETE_MULTI_FILTERS.includes(label)) {
			const description = label.replace(underscoreRe, ' ')
			return (
				<span>
					<span className="filters__text--dim">{description}:</span>
					{' '}
					<ul className="filters__summary-list">
						{value.map(({ label, id }) => (
							<li
								key={id}
								className="filters__summary-item"
							>
								{label}
							</li>
						))}
					</ul>
				</span>
			)
		} else if (AUTOCOMPLETE_SINGLE_FILTERS.includes(label)) {
			const description = label.replace(underscoreRe, ' ')
			return (
				<span>
					<span className="filters__text--dim">{description}:</span>
					{' '}
					{value.label}
				</span>
			)
		} else {
			const description = label.replace(underscoreRe, ' ')
			return (
				<span>
					<span className="filters__text--dim">{description}:</span>
					{' '}
					{formattedValue}
				</span>
			)
		}
	}

	render() {
		const { filterValues } = this.props

		return (
			<ul className="filters__summary-list">
				{Object.keys(filterValues).map(label => (
					<li key={label} className="filters__summary-item">
						{this.renderValue(label, filterValues[label])}
					</li>
				))}
			</ul>
		)
	}
}


export default FiltersList
