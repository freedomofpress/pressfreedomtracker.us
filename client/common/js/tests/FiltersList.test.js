import React from 'react';
import { shallow, mount } from 'enzyme';
import FiltersList from '~/filtering/FiltersList';
import {
    CATEGORIES_FILTER,
    FILTER_VALUES,
    BOOL_FILTER_VALUES,
    DATE_FILTER_VALUES,
    CHOICE_FILTER_VALUES,
    INT_FILTER_VALUES,
} from '~/tests/factories/CategoriesFilterFactory';

describe('FiltersList', () => {
    let categories;
    let filterValues;

    beforeEach(() => {
        categories = CATEGORIES_FILTER;
        filterValues = FILTER_VALUES;
    });

    afterEach(() => {
        categories = null;
        filterValues = null;
    });

    it('should render without error in "debug" mode', () => {
        const component = shallow(
            <FiltersList
                categories={categories}
                filterValues={filterValues}
            />
        );

        expect(component.debug()).toMatchSnapshot();
    });

    it('should render FiltersList properly with Boolean type', () => {
        const component = mount(
            <FiltersList
                categories={categories}
                filterValues={BOOL_FILTER_VALUES}
            />
        );
        const filterList = component.find('.filters__summary-item');
        expect(filterList.text()).toEqual(
            'Charged under espionage act?: Yes'
        );
    });

    it('should render FiltersList properly with Date type', () => {
        const component = mount(
            <FiltersList
                categories={categories}
                filterValues={DATE_FILTER_VALUES}
            />
        );
        const filterList = component.find('.filters__summary-item');
        expect(filterList.text()).toEqual(
            'Took place between: Sep 25th 2019 â€“ Sep 26th 2019'
        );
    });

    it('should render FiltersList properly with Choices type', () => {
        const component = mount(
            <FiltersList
                categories={categories}
                filterValues={CHOICE_FILTER_VALUES}
            />
        );
        const filterList = component.find('.filters__summary-item');
        expect(filterList.text()).toEqual(
            'Status of prior restraint: pending'
        );
    });

    it('should render FiltersList properly with Int type', () => {
        const component = mount(
            <FiltersList
                categories={categories}
                filterValues={INT_FILTER_VALUES}
            />
        );
        const filterList = component.find('.filters__summary-item');
        expect(filterList.text()).toEqual(
            `Updated in the last: ${INT_FILTER_VALUES['recently_updated']} days`
        );
    });
});
