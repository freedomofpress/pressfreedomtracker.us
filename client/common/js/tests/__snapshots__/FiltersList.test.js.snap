// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersList should render without error in "debug" mode 1`] = `
"<ul className=\\"filters__summary-list\\">
  <li className=\\"filters__summary-item\\">
    <span>
      <span className=\\"filters__text--dim\\">
        Charged under espionage act?
        :
      </span>
       
      Yes
    </span>
  </li>
</ul>"
`;
